### ISSUE TITLE
[Feature Request]: [JAM][Workzone] Support custom theme in theming service

### ISSUE THREAD

**Issue by TinaC** [SAP/ui5-webcomponents#6758]
### Feature Request Description

This is a continuation of this feature request.
https://github.com/SAP/ui5-webcomponents/issues/5730

Seems the issue is resolved with: https://github.com/SAP/ui5-webcomponents/pull/5887

But no update in the document:
https://sap.github.io/ui5-webcomponents/playground/customizing/theme/

How should I use custom theme using `setTheme`, how do I provide the theming service host to ui5 web component so it can fetch custom theme info from it?

Our SAP production WorkZone already integrated with Theming service, so we want to ui5 web component load resource directly from Theming service(just like UI5 did [Setting Themes in UI5](https://sapui5.hana.ondemand.com/sdk/#/topic/e9fc648661d84ed89360bbec3ae02611)) instead let the user download the zip file to the system. 

# `setTheme('sap_fiori_3_dark')`

<img width="996" alt="image" src="https://user-images.githubusercontent.com/6975707/226789432-64c10665-d413-4675-99ff-2949c7f2143e.png">

# `setTheme('sap_horizon')`

<img width="973" alt="image" src="https://user-images.githubusercontent.com/6975707/226789368-1f21910a-64cb-43d1-abff-1d2e154e526e.png">

# `setTheme('my_customized_horizon')`

As there is no way of providing the base theme of the custom theme, the color and SAP Icon is not right.

<img width="987" alt="image" src="https://user-images.githubusercontent.com/6975707/227492624-bb97f348-5940-49f6-9492-394e4bb25fe1.png">

### Proposed Solution

Provide an API to provide theming service host to ui5 web component, so it can know where to get the custom theme info.

This is how UI5 do it:
```js
  if (isCustomTheme(themeName)) {
    // applyTheme('my_custom_horizon_theme', 'https://themingservicehost' + '/UI5/')
    sap.ui.getCore().applyTheme(themeName, themeUrl + '/UI5/');
  } else {
    // applyTheme('sap_horizion')
    sap.ui.getCore().applyTheme(themeName);
  }
```

### Additional Context

For now, custom theme will always failed at Workzone due to this issue, we are going to release in April 10th 

### Priority

High

### Privacy Policy

- [X] I’m not disclosing any internal or sensitive information.

**Comment by s-todorova** [comment]
Hello @SAP/ui5-webcomponents-topic-core,

Could you please take a look at this feature request?

Best regards,
Siyana

**Comment by TinaC** [comment]
I have checked the source code of `applyTheme`
It seems if it is a external theme / custom theme, ui5 web component will give up on handling it, just use CSS variable files in the host system, right?
https://github.com/SAP/ui5-webcomponents/blob/main/packages/base/src/theming/applyTheme.ts#L77 

But the question is now we want to handle different version of SAP Icons in the system:
> https://wiki.one.int.sap/wiki/display/visualcore/SAP-icon+Updates 
Horizon theme uses version 5.xx icons and older themes such as Quartz and Belize or older still use version 4.xx

Since ui5 web component use SVG icons in `Icon` component instead of icon font file, so the host system (Workzone or UI5) cannot do icon version change in custom theme for UI5 web component

**Comment by nnaydenow** [comment]
Hi @TinaC,

Currently we support only custom theme roots only if they are provided by URL parameter. For example:
```
https://example.com/?sap-ui-theme=my_custom_theme@https://example-theme-provider.com
```
Providing the this information by the URL our framework will load my_custom_theme CSS resources from https://example-theme-provider.com.

Determining which icons to be used is based on theme origin. For example if your custom theme extends SAP Quartz version 5.xx of icons will be used and if your custom theme extends SAP Belize version 4.xx of icons will be used

**Comment by TinaC** [comment]
Hi @nnaydenow,

As a complex application, our production use [redirect_to](https://api.rubyonrails.org/v7.0.4.2/classes/ActionController/Redirecting.html) widely, which will lose the URL params,

Our solution is save the URL params in the session storage, and call `setTheme()` with theme info in the session storage after the new page load.

So we want `setTheme(themingServiceHost)` to support pass the theme host as the second params(just like [UI5 applyTheme() did](https://sapui5.hana.ondemand.com/sdk/#/topic/e9fc648661d84ed89360bbec3ae02611)), something like `setTheme(themeName, themingServiceHost)`

> Determining which icons to be used is based on theme origin. For example if your custom theme extends SAP Horizon version 5.xx of icons will be used and if your custom theme extends any other theme version 4.xx of icons will be used

This is exactly what we want, but since URL params will lost in redirect, this behavior will fail sometimes.

**Comment by ilhan007** [comment]
Hello @TinaC could you try the following:
```ts
setTheme("workzone_horizon"); 
setThemeRoot("http://example.com"); // not public yet, but we will introduce something similar, either extend setTheme as suggested, or another convenient API
```

BR,
ilhan

**Comment by TinaC** [comment]
Hi @ilhan007 , 

thanks for providing this workaround.
https://github.com/SAP/ui5-webcomponents/blame/main/packages/base/src/config/ThemeRoots.ts
This function is only available in @ui5/webcomponents v1.11.0, right?
Currently we are in v1.6.0.
I will have a try after we upgrade the version.

BTW, about set custom theme in URL:
> https://example.com/?sap-ui-theme=my_custom_theme@https://example-theme-provider.com

Could you update it in the document:
https://sap.github.io/ui5-webcomponents/playground/customizing/theme/

Thanks,
Tina

**Comment by nnaydenow** [comment]
Hi @TinaC,

The documentation is updated and you can check it in the page you've provided: https://sap.github.io/ui5-webcomponents/playground/customizing/theme/

**Comment by TinaC** [comment]
Hi @ilhan007 , 
I saw you add a 1.13.0 label, I'm trying to upgrade to the latest 1.12.0, is there other fixes in 1.13.0 version?
When will 1.13.0 release? Let me check if I can catch the dev close date of this release.

**Comment by ilhan007** [comment]
Hello @TinaC 
- we updated the Documentation https://sap.github.io/ui5-webcomponents/playground/customizing/theme/
- proposed solution with the use of "setThemeRoot".

Yes, it's sufficient to upgrade to 1.11, 1.12 is also fine.

We have moved the item to 1.13, because we need more time to decide on the public API, to make setThemeRoot public (as it's currently private) or to enhance "setTheme" to support a second parameter for theme root, there are also other options.
1.13 will be produced in the 30 April - 4 May period. If we don't come up with a final API, we might move the item for the next release.
After we decide on the API, merge and release it, you will need to adopt it when upgrading in future, but it won't be a big change - to add a second parameter, to replace one method with another, something along these lines.

Here is the release plan of the UI5 Web Components:
<img width="898" alt="Screenshot 2023-04-26 at 10 41 30" src="https://user-images.githubusercontent.com/15702139/234504678-1ddd95a4-2ba4-46b9-809b-2a0d545c9eb4.png">
Described at the bottom of the following [page](https://github.com/SAP/ui5-webcomponents/projects?type=classic).
